### PR TITLE
krainstance: set correct issuer DN in uid=ipakra entry

### PR DESCRIPTION
**Pull request by frasertweedale**
If IPA CA has custom subject DN (not "CN=Certificate
Authority,{subject_base}"), the uid=ipakra people entry gets an
incorrect 'description' attribute.  The issuer DN in the
'description' attribute is based on the aforementioned pattern,
instead of the actual IPA CA subject DN.

Update KRAInstance.configure_instance() to require the CA subject DN
argument.  Update ipaserver.install.kra.install() to pass the CA
subject DN.

Fixes: https://pagure.io/freeipa/issue/8084

**Verification steps**: https://bugzilla.redhat.com/show_bug.cgi?id=1758404#c2

### CODE REVIEW

**Comment by flo-renaud** [comment]
Hi @frasertweedale 
thanks for the PR, works for me.
I see that we don't have any test with custom subject. Could you add a new test that calls ipa-server-install with `--subject-base` and `--ca-subject`, and ensures that
- the certs in /etc/pki/pki-tomcat/alias have the right subject
- the attribute ipaCertificateSubjectBase of cn=ipaConfig,cn=etc,$BASEDN contains the right val
- the attribute description of uid=pkidbuser,ou=people,o=ipaca and uid=ipara,ou=people,o=ipaca contain the right val?

**Comment by frasertweedale** [comment]
On Mon, Oct 07, 2019 at 02:04:30AM -0700, Florence Blanc-Renaud wrote:
> Hi @frasertweedale 
> thanks for the PR, works for me.
> I see that we don't have any test with custom subject. Could you add a new test that calls ipa-server-install with `--subject-base` and `--ca-subject`, and ensures that
> - the certs in /etc/pki/pki-tomcat/alias have the right subject
> - the attribute ipaCertificateSubjectBase of cn=ipaConfig,cn=etc,$BASEDN contains the right val
> - the attribute description of uid=pkidbuser,ou=people,o=ipaca and uid=ipara,ou=people,o=ipaca contain the right val?
> 
Flo, I have in mind (but have not yet created tickets) to
generate random-ish Subject DN in tests, some percentage of the
time, so that we can find more corner cases.

I suppose for now it can suffice to have a test that uses a fixed or
deterministic "custom" subject DN.

Thanks for reviewing!

**Comment by flo-renaud** [comment]
Hi @frasertweedale 

> I suppose for now it can suffice to have a test that uses a fixed or
> deterministic "custom" subject DN.
Yes, minimal checks would suffice for now.

One additional question that comes to my mind: we could also add an upgrade plugin to fix the "description" attribute for already installed configs with the bug. What do you think about it?

**Comment by frasertweedale** [comment]
@flo-renaud yes, upgrade script for repair is a good idea, I will do that too :)  Adding WIP label.

**Comment by frasertweedale** [comment]
Added upgrade script; still need to write the rudimentary tests.

**Comment by frasertweedale** [comment]
Successful CI run of Custom Subject DN tests via temp commit: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/ab857938-ee34-11e9-abf3-fa163e2e23e6/

**Comment by frasertweedale** [comment]
@flo-renaud thanks for review.  I moved the upgrade behaviour to an Upgrader plugin and added the requested CI job definitions.  Link to CI results of the temp commit is in comment https://github.com/freeipa/freeipa/pull/3764#issuecomment-541498476.

File: ipatests/prci_definitions/nightly_master_pki.yaml [hunk 1]
```diff
@@ -782,3 +782,15 @@ jobs:
...
         topology: *master_1repl
+
+  fedora-30/test_ca_custom_sdn:
+    requires: [fedora-30/build]
+    priority: 50
+    job:
+      class: RunPytest
+      args:
+        build_url: '{fedora-30/build_url}'
+        test_suite: test_integration/test_ca_custom_sdn.py
```
Comment (flo-renaud) [inline ipatests/prci_definitions/nightly_master_pki.yaml]: We need to add `update_packages: True` in nightly_master_pki.yaml, so that the latest build of PKI is installed.

**Comment by frasertweedale** [comment]
Thanks @flo-renaud, fixed.

**Comment by flo-renaud** [comment]
Hi @frasertweedale 
the test failure (test_smb) is linked to an infra issue, and PR  #3792 will temporarily remove the test from gating. Please wait for #3792 to be merged, and rebase, then the CI should pass.
Apart from that, LGTM.

**Comment by flo-renaud** [comment]
master:

* 326d417d98b092e175623cd28e46586df00e60a2 krainstance: set correct issuer DN in uid=ipakra entry
* 7ea50ff76d2ee5367b75d999c2baa2a7a480f34a upgrade: fix ipakra people entry 'description' attribute
* e767386e7120be3515d6a34529b51ae658248038 test_integration: add tests for custom CA subject DN

**Comment by flo-renaud** [comment]
Hi @frasertweedale 
the automated backport failed, can you manually backport to ipa-4-6, ipa-4-7 and ipa-4-8 branches? Thanks
